service/history/retry.go

```diff
@@ -117,8 +117,8 @@ func isRetryable(failure *failurepb.Failure, nonRetryableTypes []string) bool {
 }
 
 func getCauseFailure(failure *failurepb.Failure) *failurepb.Failure {
-	for ; failure.GetCause() != nil; failure = failure.GetCause() {
+	// Unwrap failures till the first ApplicationFailure because only first ApplicationFailure controls retryable.
```

Comment:
I don't think we want to unwrap up to the first application failure. We want to get immediate cause and if it is not an application failure then stop.

Reply:
I rethink this logic myself: we will unwrap only `ChildWorkflowExectutionFailure` and `ActivityFailure`.